src/converter/jfr2pprof.java

```diff
@@ -0,0 +1,243 @@
+    public static final int VALUETYPE_UNIT = 2;
+
+    // Sample IDs
+    public static final int SAMPLE_LOCATION_ID = 1;
+    public static final int SAMPLE_VALUE = 2;
+
+    // Location IDs
+    public static final int LOCATION_ID = 1;
+    public static final int LOCATION_LINE = 4;
+
+    // Line IDs
+    public static final int LINE_FUNCTION_ID = 1;
+
+    // Function IDs
+    public static final int FUNCTION_ID = 1;
+    public static final int FUNCTION_NAME = 2;
+
+    // `Proto` instances are mutable, careful with reordering
+    public void dump(final FileOutputStream out) throws Exception {
+        // Mutable IDs, need to start at 1
+        int functionId = 1;
+        int locationId = 1;
+        int stringId = 1;
+
+        // Used to de-dupe
+        final Map<Method, Long> functions = new HashMap<>();
+        final Map<Method, Long> locations = new HashMap<>();
+
+        final List<ExecutionSample> samples = reader.readAllEvents(ExecutionSample.class);
+        final Dictionary<StackTrace> stackTraces = reader.stackTraces;
+        final long startTicks = reader.startTicks;
+
+        final Proto profile = new Proto(200_000)
+            .field(PROFILE_TIME_NANOS, reader.startNanos)
+            .field(PROFILE_DURATION_NANOS, reader.durationNanos())
+            .field(PROFILE_DEFAULT_SAMPLE_TYPE, 0L)
+            .field(PROFILE_STRING_TABLE, "".getBytes(StandardCharsets.UTF_8)) // "" needs to be index 0
+            .field(PROFILE_STRING_TABLE, "Async Profiler".getBytes(StandardCharsets.UTF_8))
+            .field(PROFILE_COMMENT, stringId++);
+
+        // Sample Type
+        final String sampleTypeName = "cpu";
+        final String sampleTypeUnit = "nanoseconds";
+
+        final Proto sampleType = new Proto(10_000);
+
+        profile.field(PROFILE_STRING_TABLE, sampleTypeName.getBytes(StandardCharsets.UTF_8));
+        sampleType.field(VALUETYPE_TYPE, stringId++);
+
+        profile.field(PROFILE_STRING_TABLE, sampleTypeUnit.getBytes(StandardCharsets.UTF_8));
+        sampleType.field(VALUETYPE_UNIT, stringId++);
+
+        profile.field(PROFILE_SAMPLE_TYPE, sampleType);
+
+        long previousTime = startTicks; // Mutate this to keep track of time deltas
+
+        // Iterate over samples
+        for (final ExecutionSample jfrSample : samples) {
+            final int sampleThreadState = jfrSample.threadState;
+            final long sampleTime = jfrSample.time;
+            final int sampleThreadId = jfrSample.tid;
+            final int sampleStackTraceId = jfrSample.stackTraceId;
+            final StackTrace stackTrace = stackTraces.get(sampleStackTraceId);
+            final long[] methods = stackTrace.methods;
+            final byte[] types = stackTrace.types;
+
+            // Need to reverse the order of the elements; index 0 must be the leaf node
+            int current = 0;
+
+            final double nanosSinceLastSample = (jfrSample.time - previousTime) * 1_000_000_000 / reader.ticksPerSec;
```

Comment:
The result of `long / long` is stored into `double`

Reply:
I guess we want to round it into a `long` anyways...